### ISSUE TITLE
dealing with many different but related graphs

### ISSUE THREAD

**Issue by iceland20** [pyg-team/pytorch_geometric#1947]
Hi,

I have many graphs that can be indexed by time (weeks),  they represent the same social network. I want to predict some of the features for each of those recorded graphs.

This earlier [issue](https://github.com/rusty1s/pytorch_geometric/issues/817) helps somewhat. However, in my case not only feature values change, but also nodes in x and edges in edge_index do. Following the suggestion there
```
data1 = Data(edge_index=edge_index1, x=x1)
data2 = Data(edge_index=edge_index2, x=x2)
batch = Batch.from_data_list([data1, data2])
```
is batching all these graphs one on top of another the preferred/only way? I have a lot of them... Or is there some generic structure I can use and reuse?

More importantly, given that nodes and edges change from one graph to another, can I be certain that training will not be effected by nodes (and edges) moving around, changing their order, total number and so on? (To be sure, I set up the Net class generically as if I deal with just one recorded graph that will be summarized in some dataset `Data(edge_index=edge_index, x=x)`.)

Thank you very much for your time.

**Comment by iceland20** [comment]
I have some additional questions on the same topic.

My graph data structure is a list dataAll=[data_1, data_2, data_3, ... data_525], with each data_i having varying nodes and edges. (I tried to use `batch = Batch.from_data_list([data_1, data_2, ..., data_525])`, but that just merges all my data subsets together, which i don't think is correct, if anything for memory load.)

When defining the network, I added `node_dim=1` to let the message passing algorithm understand this.
```
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        nnA = torch.nn.Sequential(torch.nn.Linear(data.num_edge_features, data.num_node_features*num_n1, node_dim=1))
        self.conv = NNConv(data.num_node_features, num_n1, nnA, aggr='mean', node_dim=1)
```
with the forward function defined without such modifications.

However, when I'm trying to send the data to CUDA with `data = dataAll.to(device)`, it does not work for lists.
```
device = torch.device('cuda')
data = dataAll.to(device)
model = Net().to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=2e-5)
```

Also, I am not sure how to set up the actual training, given this weird data structure. Is it something like below, or do I need to use batches (if the latter, as far as I understand it's not going to be standard mini-batching because my data is quite rigidly structured)?
```
model.train()
for epoch in range(100):
    optimizer.zero_grad()
    out = model(data_optimize)
    #calculate loss and so on
```
Lastly, where and how do I define training and testing masks, given that they are different for subsets data_i in dataAll structure?

I would be really grateful for some constructive advice.

**Comment by rusty1s** [comment]
As you correctly said, batching *all* graphs together will usually not work due to memory limitations. I'm not sure what your training objective looks like, but you should consider training in mini-batches (without `node_dim=1` since you are not operating on static graphs). For a temporal network (where examples in a batch should be consecutive), this could look like:
```python
loader = torch.utils.DataLoader(range(num_examples - batch_size), batch_size=1, shuffle=True)
for idx in loader:
    batch = Batch.from_data_list(dataset[idx:idx+batch_size)
```
Hope that this is useful to you! This should work even when nodes are changing their order.
In case examples inside a batch do not need to be consecutive, you can simply use the PyG dataloader though:
```python
from torch_geometric.data import DataLoader
loader = DataLoader(dataset, batch_size=32)
```

**Comment by iceland20** [comment]
Thank you, @rusty1s, definitely useful!

All right, so I got confused about `node_dim=1`, and I need your consecutive DataLoader formulation. Here, my training objective is node features.

This is what I get in the end, is it what you had in mind?
```
device = torch.device('cuda')

num_ep = 100
num_examples = 525
batch_size = 1
loader = torch.utils.data.DataLoader(range(num_examples-batch_size), batch_size=1, shuffle=True)

for epoch in range(num_ep):
    for idx in loader:
        batch = Batch.from_data_list(dataAll[idx:idx+batch_size])
        
        dataAll[idx].train_mask = torch.tensor(train_set[idx, :, 0], dtype=torch.bool)
        dataAll[idx].test_mask = torch.tensor(test_set[idx, :, 0], dtype=torch.bool)

        data = dataAll[idx].to(device)
        model = Net().to(device)
        optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=2e-5)
        model.train()

        optimizer.zero_grad()
        out = model(data)[data.train_mask]
        trg = ( data.y[data.train_mask,0] )
        loss = torch.nn.functional.l1_loss( out, trg )
        loss.backward()
        optimizer.step()
```

One additional clarification: given (1) such batching and training procedures, and that (2) my `Net` class in its `__init__(self)` definition, as you can see from my previous posting, refers to input as `data` (where `data = dataAll[idx].to(device)` in the training module) and contains a random initialization of network's parameters `self.B = torch.nn.Parameter(torch.randn(num_out_features))`, will the learning process be continuous? I'm worried about it being restarted every time I change batch and/or epoch.

Thank you again!

**Comment by rusty1s** [comment]
It currently looks like you input a single example into your model, and do not make use of batch information (in particular, because `batch` is never used). It currently trains the model in a static fashion (one example in -> one prediction out), which I do believe is not what you are trying to achieve. You have two alternatives:
1. Use `batch` in `model`
2. Apply backpropagation trough time by an additional for-loop
```python
for idx in loader:
    optimizer.zero_grad()
    for b in range(batch_size):
        out = model(dataAll[idx + b]
        loss += ...
    loss.backward()
   optimizer.step()
```
Nonetheless, your model will not take temporal information into account as long as you do not design it to do so (e.g. via a memory module).

**Comment by iceland20** [comment]
Hmm, I thought my formulation was actually doing something like what your second alternative says. But you are right, the batch was never used and was redundant. Could you please take a look at another attempt?

Given the following Net class (same as above):
```
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        nnA = torch.nn.Sequential(torch.nn.Linear(data.num_edge_features, data.num_node_features*num_n1, node_dim=1))
        self.conv = NNConv(data.num_node_features, num_n1, nnA, aggr='mean', node_dim=1)
        #subsequent layers including
        self.B = torch.nn.Parameter(torch.randn(num_out_features))
        #and so on
```

I am now explicitly using `batch` rather than `dataAll` while constantly redefining `data` that the `Net` formulation relies on:
```
device = torch.device('cuda')

num_ep = 100
num_examples = 525
batch_size = 1
loader = torch.utils.data.DataLoader(range(num_examples-batch_size), batch_size=1, shuffle=True)

for epoch in range(num_ep):
    for idx in loader:
        batch = Batch.from_data_list(dataAll[idx:idx+batch_size])
        
        batch.train_mask = torch.tensor(train_set[idx, :, 0], dtype=torch.bool)
        batch.test_mask = torch.tensor(test_set[idx, :, 0], dtype=torch.bool)

        data = batch.to(device)
        model = Net().to(device)
        optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=2e-5)
        model.train()

        optimizer.zero_grad()
        out = model(data)[data.train_mask]
        trg = ( data.y[data.train_mask,0] )
        loss = torch.nn.functional.l1_loss( out, trg )
        loss.backward()
        optimizer.step()
```

Understand about the memory module or something like that, for the time being I just want to make sure the plain vanilla formulation is set up correctly. So I don't mind if I feed example graphs one by one as long as after using all `num_examples` that the `loader` supplies, my network parameters are optimized with respect to all the data I have (in each epoch, that is).

Thank you so much for your help, really appreciate it!

**Comment by rusty1s** [comment]
That looks mostly correct to me, although it seems like the dimensions of `model(data)` (the output of every node in the batch) and `data.train_mask` (the mask for the first graph in the batch) do not match.

**Comment by iceland20** [comment]
Thank you! This is interesting. My `train_mask` is graph-specific and has `num_nodes` masks. At the same time, my Net returns `num_out_features` of predictions, so as far as I understand `model(data)` is returning `[num_nodes, num_out_features]` of them. In the case considered, every batch is one graph (`dataAll[idx]` contains exactly one graph), so the dimensions should match.

What am I  missing? (When I run the code, it does not give an error, but of course this does not mean the code is correct.)

**Comment by rusty1s** [comment]
Yes, that only works for batch_sizes of 1. In case of larger batch sizes, you need to take a slice of the `train_set` for masks: `train_set[idx:idx+batch_size]`.

**Comment by iceland20** [comment]
Right, I was thinking solely about batches of 1. Thanks for being pedantic and spotting the potential error!

Very good, thank you very much for your guidance on the problem I raised, @rusty1s! Extremely grateful for your inputs.

**Comment by iceland20** [comment]
Returning to this issue again, I'm having problems with learning: when I run the above procedure, it does seem to optimize and produce output (interim and final losses), but the loss function just moves around without much improvement. Even though I know that my data have a lot of signal. When raising this issue, I was [concerned](https://github.com/rusty1s/pytorch_geometric/issues/1947#issuecomment-751851919) whether the above setup ensures that parameters are not reset randomly from one batch to another and from one epoch to another. It seems like this might be happening now.

Is there some function I need to define or parameter to set that I might have missed? And what would be the best way of tracking whether learning is "continuous" (as opposed to restarting every bacth/epoch)?

Thank you!

**Comment by rusty1s** [comment]
That is correct. You need to move the initialization of the model and the optimizer outside the training loop.

**Comment by iceland20** [comment]
Thank you, @rusty1s! Actually, I tried to do that some time ago, but the reason I moved it inside the loop is because I need to define `data` that my `Net()` depends on. But now when you said it, I tried again.

So, I moved
```
model = Net().to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=2e-5)
```
outside the training loop. This time, before defining `model`, I initialize `data` as the first batch of the dataset:
`data = Batch.from_data_list(dataAll[0:0+batch_size])`
with the intention of sending new batch to GPU every `idx` iteration using `data = batch.to(device)` (same as before).

The results looks not exactly correct, loss function improves a bit and then stalls, so it seems like learning is only happening with respect to that very first batch of data. Basically, my question is how to define `data` for `Net()` to use given that my learning happens batch by batch and so I am not sending all the data to GPU at once? (I tried to look at the docs, but those I found consider batching without showing the training part, so the issue never arises.)

**Comment by rusty1s** [comment]
I'm not really sure I understand. You model should never be dependent on a specific data object, only based on same attributes of it (e.g., `num_nodes`, `num_features`, ...).

**Comment by iceland20** [comment]
I understand this, but here where my problem is. My collection of graphs is a list `dataAll`. When I'm [defining](https://github.com/rusty1s/pytorch_geometric/issues/1947#issuecomment-752612057) the graph `Net()` class I refer to `data` as input, for example `self.conv = NNConv(data.num_node_features, num_n1, nnA, aggr='mean', node_dim=1)`. Then I am changing this `data` object inside the training loop as `data = batch.to(device)`, slicing the list piece by piece as `batch = Batch.from_data_list(dataAll[idx:idx+batch_size])`. However, when defining the `Net()`, I must have the `data` object initialized somehow.
1. I tried to use the first element of `dataAll`, that is `data = Batch.from_data_list(dataAll[0:0+batch_size])`. This does not give an error, but results seem wrong, I guess because once I initialize `data` this way, my `Net()` takes a specific object rather than a placeholder, so it uses only the first graph for learning.
2. I also tried to define it generically in an obvious way like `data=torch_geometric.data.Data()`, but `model = Net().to(device)` returns an error then.

Could you please suggest how to initialize `data` in such a way that I could define `Net()`, but then also could supply new concrete inputs from each batch?

The batching examples I've seen so far have one huge graph that they supply to the network class, but which is then split into batches while learning, so my issue never arises there.

Sorry for abusing your patience, @rusty1s! And thank you so much for your time!

**Comment by rusty1s** [comment]
I see. That is not a problem. You can define your model that way. A more elegant solution would be to pass the number of input channels as an argument to `Net`.

Nonetheless, training will take place on all data inputs you pass into the model. As such, your code looks correct to me.

**Comment by iceland20** [comment]
Thank you, but my question remains open: how do I initialize `data` object in such a way that I could define `Net` using `data` as a placeholder for input graph (and then `model` could accept this `Net`), allowing me to use batching and change the concrete graph `data` refers to?

**Comment by rusty1s** [comment]
Yes, that's fine since the number of features do not change across different data objects.

**Comment by iceland20** [comment]
Well, it definitely makes sense intuitively, glad that this is also the case in the way PyG works in reality! Okay, will play a bit more with my model to see what goes wrong.

Thank you so much for the help!
